change to use std::thread

yumetodo opened the pull request:
use `std::thread`.
# benchmark

```
960*720
SDeContrast
Midtone : 50
Strength: 6
```

```
max:最大値
min:最小値
sum:合計
avg:平均
stdev:母集団標準偏差
se:標準誤差
95%CL:95%信頼区間
C.I.max:95%信頼区間を平均に加えたもの
C.I.min:95%信頼区間から平均を引いたもの
```

|  | before[ns] | after[ns] |
| --- | --- | --- |
| max | 70144353 | 90190587 |
| min | 25787378 | 22779376 |
| avg. | 38799114.8973277 | 36900188.2022989 |
| count | 2133 | 2175 |
| stdev | 4120405.38657474 | 8409477.70072626 |
| se | 89216.3637673064 | 180318.129087751 |
| 95%CL | 174860.859815545 | 353417.038771636 |
| C.I.max | 38973975.7571433 | 37253605.2410705 |
| C.I.min | 38624254.0375122 | 36546771.1635272 |

yumetodo committed:
change to use std::thread

MaverickTse commented:
Sorry, nope. The change is too big for such marginal improvement. If there are 10ms difference on 720P, then I'll consider.

MaverickTse commented:
Also you need to turn off the /Qpar option in the project file, otherwise the results cannot be compared.

yumetodo committed:
disable  /Qpar

MaverickTse commented:
show me the new benchmark plz

yumetodo commented:
@MaverickTse prease wait. I'm working #10 to investigate this.

yumetodo committed:
Merge remote-tracking branch 'refs/remotes/upstream/master'

Conflicts:
	SigColorFastAviUtl/sigcolorfastaviutl.cpp

MaverickTse commented:
I'd suggest against working on this further more. I have checked on my computer that /Qpar works correctly(using 12 out of my 13 cpu cores) and now I use a simple float[4] array on stack which posed no problem on row-wise parallelization.
The use of std::thread really makes the code a lot more complex and I have no confident I can maintain that.

yumetodo committed:
Merge remote-tracking branch 'refs/remotes/upstream/master'

Conflicts:
	SigColorFastAviUtl/sigcolorfastaviutl.cpp

yumetodo commented:
|          | before[ns]       | after[ns]        |
|----------|------------------|------------------|
| max      | 58851592         | 122142375        |
| min      | 15095044         | 17856938         |
| avg.     | 40697668.4706525 | 39822508.8662606 |
| count    | 3663             | 3776             |
| stdev    | 5228744.19346652 | 8064316.19743504 |
| se       | 86393.0757542369 | 131235.567943636 |
| 95%CL    | 169327.316991945 | 257216.986660185 |
| C.L.max  | 40866995.7876444 | 40079725.8529208 |
| C.L.min  | 40528341.1536605 | 39565291.8796004 |

very small speed-up...

yumetodo committed:
improve Job sharing

MaverickTse commented:
The average is only 1ms difference while stdev is a LOT wider... so nope. 
I have also tried the Concurrency library available in VS2015 (ppl.h), also to no avail. The /Qpar is still the most stable and fastest.

If you want to continue to play with the multi-threading issue, try to parallelize the LUT lookup. So far I failed to do this and time of processing increase with number of channels being processed.

yumetodo committed:
Merge branch 'master'

Conflicts:
	SigColorFastAviUtl/sigcolorfastaviutl.cpp

yumetodo committed:
specify noexcept

yumetodo committed:
change charcode to UTF-8 with BOM

yumetodo committed:
make new header for `parallel::per_for`

yumetodo committed:
add aync_for

MaverickTse commented:
Note that I'm not going to merge unless there are enough improvement in speed.

yumetodo committed:
fix compile error

yumetodo committed:
rewrite SigmoidTable  class

stop using std::map

yumetodo committed:
remove unused cpp from project

yumetodo committed:
SigmoidTable recalc : use MT

yumetodo committed:
rewrite RSigmoidTable

yumetodo committed:
add test framework(iutest)

yumetodo committed:
add test case

yumetodo committed:
Revert "fix bug(SigmoidTable) : fix table generate range"

This reverts commit 1fde83eaf857436c1ae5c67fa3b7af336853c385.

yumetodo committed:
fix bug(SigmoidTable) : fix table generate range

yumetodo committed:
_mm_div_ps is 1%(+-2.4% N=19)faster than m128_rcp_22bit_ps

avg	11352.5789473684	11233.5263157895
sd	314.719811192093	307.312693217743
sc	141.512685461962	138.182100227707
sc.max	11494.0916328304	11371.7084160172
sc.min	11211.0662619065	11095.3442155618
	m128_rcp_22bit_ps	_mm_div_ps

yumetodo committed:
fix testcase

yumetodo committed:
fix testcase

yumetodo committed:
fix bug(RSigmoidTable): avoid multiple insertaion

https://twitter.com/yumetodo/status/797363830200242176
clear table before creating and check the place is not 0 where try to insert to.

yumetodo commented:
やっとテーブルの動的メモリー確保を削れた。
テストコードより、もとの計算結果と100%一致することを確認。

これでやっとAvUtlでの実地試験に望める。

yumetodo commented:
Ut Video Codec SuiteなAVI素材だとなんかバグっているけど
![image](https://cloud.githubusercontent.com/assets/10869046/20237427/7e8a6c5c-a915-11e6-9048-a66db961399d.png)
とりあえず1000 micro sec.～3000 micro sec.にまで高速化できたので成果はあった。

MaverickTse commented:
このバッグを治ったらmerge する

yumetodo committed:
add testcase for out_of_range

yumetodo committed:
fix bug(SigmoidTable): fix when (input < 0 || 4096 < input)

yumetodo committed:
partial fix bug(RSigmoidTable): fix when input < 0

yumetodo committed:
refactor(SigmoidTable): change some constant to private

yumetodo committed:
fix bug(RSigmoidTable): fix when 4096 < input

yumetodo commented:
とりあえず範囲外についても元と同じ挙動にした。これよりAviUtl上でのテストに入る

yumetodo commented:
治ったっぽい・・？
![image](https://cloud.githubusercontent.com/assets/10869046/20237670/d23349f8-a91c-11e6-932e-d276e3e6372b.png)

yumetodo committed:
benchmark: change time units(milli sec. -> micro sec.)

yumetodo commented on SigColorFastAviUtl/RSigmoidTable.h:
```diff
@@ -45,11 +45,14 @@ inline void RSigmoidTable::change_param(float midtone, float strength) noexcept
 		for (value_type i = pre + 1; i <= x; ++i) this->insert_to_table(i, y);//fill blanc and insert new value
 		pre = x;
 	}
+	on_error_value_high_ = table_[pre];
```
Sigmoid関数の性質上、入力値が単調増加すれば出力も単調増加する。これは逆関数でも適用できる。

書き換え前のコードでは入力が4097以上のとき`std::map::rbegin()`が指す要素を返していた。
これは`x`の最大値に対応する`y`であり、上述したSigmoid関数の性質より、最大値は最後に計算で得た`x`に対応する`y`である。

よって、その値を`on_error_value_high_`に格納する

MaverickTse replied:
つまりSpecial caseで対応するか…できればもっとelegantな対応が欲しい。
ま、正確に動ける以上俺は構わんけど
